### ISSUE TITLE
very basic questions

### ISSUE THREAD

**Issue by ag1988** [ptillet/torch-blocksparse#14]
Hi, I am tying to use torch_blocksparse.Linear(). The call to MatMul inside its init method needs to be updated with the new API. Also I just want to create a block-diagonal weights matrix of dim 4x4 with block size 2 where the 2 non-diagonal blocks are 0 and the 2 diagonal blocks have trainable weights. Can you tell me how to create this using torch_blocksparse.Linear()? I am lost in the layout dimension. I defined it as 
W = Linear(4, 4, 2, torch.BoolTensor([[1,0], [0, 1]])).cuda()
x = torch.ones((1, 4))
W(x) gives error.
PS what does sparse dimension mean? Sorry for so many questions - this library is very useful for my purpose but I am struggling to understand the meaning of the function args :cry: .
Also after a while the entire library gets stuck due to some error with lock files - i dont know what those are.
![image](https://user-images.githubusercontent.com/34059288/87127787-31d34100-c297-11ea-8ed1-84d236caf45a.png)

**Comment by ptillet** [comment]
Hi!

Thanks for your interest in this project. I'm sorry that this is still rather experimental, but I will try to answer all your questions.

The  sparsity layout works as follows. Assume that you have a batched matrix multiplication, say C = AB where A has shape (H, M, K) and B has shape (H, K, N). The layout would be 3D tensor that has, in the case of DSD for example, shape (H, M//block, K//block). The motivation for not blocking H is to allow for more flexible layouts in sparse self-attention. Now, in some cases you may want this layout to be shared for a bunch of batches Z. In this case, the shape of A would be (Z, H, M, K) and the shape of B would be (Z, H, K, N), but for DSD the layout would still have shape (H, M//block, K//block).

One important limitation of torch-blocksparse is that it only supports block sizes 8, 16, 32, 64. Sorry for not making this more explicit; I'll add error message.

As for the locking, this is a bug that I thought I had resolved. You may want to reinstall Triton manually:  pip install -e "git+https://github.com/ptillet/triton.git#egg=triton&subdirectory=python". Seems like the version on pip does not include this bugfix.

**Comment by ag1988** [comment]
Hi Philippe,

Thank you so much for the clarifications and for taking the time out to answer my questions! They're really helpful - now I understand what exactly "block" is supposed to mean. The good thing is that after your clarifications I can also read the code and get even more clarity. I'll re-install Triton manually and try again. 
All the very best with your thesis :+1: 

Thanks again,
Ankit

**Comment by ptillet** [comment]
Cool!

Just one thing I forgot to mention. After you reinstall Triton, remember to delete the cache folder $HOME/.triton

**Comment by ag1988** [comment]
Will do. I reinstalled Triton but the error persisted so I ended up deleting /tmp/torch_extensions/ . It was fine after that. If this happens again I'll delete $HOME/.triton . Thanks again for helping me out.

**Comment by ptillet** [comment]
Cool. Please let me know if you see this error happening again. It is something that I really want to get to the bottom of and fix if it reappears.

**Comment by ag1988** [comment]
Sure, I'll let you know if this issue resurfaces. Another weird thing I keep encountering is the illegal memory access error. A piece of code will be working fine - the moment I run it again it just throws this error. When I close the notebook and restart the kernel the error persists for a while and then the same command works fine again. Anyways, you can have a look once you have more time.
![image](https://user-images.githubusercontent.com/34059288/87238971-26585500-c412-11ea-85ad-99b2349f8966.png)

**Comment by ptillet** [comment]
Can you share the code that you're running?

**Comment by ag1988** [comment]
It gives an  error only if someone else is also using the same GPU. The the code runs fine otherwise.

import torch
import torch_blocksparse

# Z: non-sparse batch dimension
# H: sparse batch dimension
# M: row dimension
# N: column dimension
Z, H, M, N, K = 1, 1, 1024, 256, 512
a = torch.rand((Z, H, M, K), dtype=torch.float32).cuda()
b = torch.rand((H, K, N), dtype=torch.float32).cuda()
# create layout for the sparse argument
block = 16
layout = torch.ones(H, K//block, N//block).long()
# create object for Dense = Dense x Sparse (dds)
# some overhead there as it pre-computes look-up tables 
# internally needed by GPU kernels
dot = torch_blocksparse.MatMul(layout, block, 'dds', trans_a=False, trans_b=False)
c = dot(a, b)
print(c.shape)

PS. The reason I asked you about the sparse layout was that in some [papers](https://arxiv.org/pdf/2005.14469.pdf) people post their performance on randomly sparse layouts. In deep learning, usually the layouts have a specific structure like block diagonal matrices, band matrices, dilations, etc - I mean obviously such a structure can be exploited for fast matrix multiplication algorithms but I am not sure if this is also the case for GPU based implementations - I dont know much about this topic but its reasonable to assume that more structure should lead to faster algorithms.

**Comment by ptillet** [comment]
Ah that's interesting. I was not aware of this bug. I assume Pytorch works when sharing GPUs though, correct? If yes, then I probably should create a separate issue for this bug specifically.

I think I see how statically predictable sparsity storage could improve the performance of unstructured sparsity. For block-sparsity, I am not sure the kernel would have to change since the amount of work done by each block is similar. That said, other aspects of the code could change (mostly launch order of the work-groups) to increase load balancing or cache performance. Regardless, do raise an issue if you see performance much better on some layouts than other at equal sparsity ratio :)

**Comment by ag1988** [comment]
Yes, Pytorch still works fine even if others are using the same device. Sure, I'll open an issue separately.

Thanks for letting me know opinion - its very helpful. Well yeah I agree if you're designing an algorithm that should work for **all** matrices with 50% sparsity then there isn't much one can do. This is also explains why the performance gains in the benchmark plots are super-significant at 50% sparsity. But on the other hand if you're designing the algorithm for only NxN matrices that are block diagonal of the form [[B11, 0], [0, B22]] where B11, B22 are dense then overall sparsity is still 50% but one can simply process the 2 dense blocks independently on much shorter vectors. Anyways, you can think about this later when you have more time :)
